### PR TITLE
PoC BOM for Play

### PR DESCRIPTION
**Pull request by ignasi35**
see also https://github.com/akka/akka/pull/29839

### CODE REVIEW

**Comment by ignasi35** [comment]
Here's a preview of what's produced:

<details>
<summary>sample</summary>

```xml
<?xml version='1.0' encoding='UTF-8'?>
<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0">
    <modelVersion>4.0.0</modelVersion>
    <groupId>com.typesafe.play</groupId>
    <artifactId>play-maven-dependencies</artifactId>
    <packaging>pom</packaging>
    <description>play-maven-dependencies</description>
    <url>https://playframework.com</url>
    <version>2.8.1+749-78742811</version>
    <licenses>
        <license>
            <name>Apache-2.0</name>
            <url>http://www.apache.org/licenses/LICENSE-2.0.html</url>
            <distribution>repo</distribution>
        </license>
    </licenses>
    <name>play-maven-dependencies</name>
    <organization>
        <name>com.typesafe.play</name>
        <url>https://playframework.com</url>
    </organization>
    <scm>
        <url>https://github.com/playframework/playframework</url>
        <connection>scm:git:git@github.com:playframework/playframework.git</connection>
    </scm>
    <developers>
        <developer>
            <id>playframework</id>
            <name>Play Framework Team</name>
            <url>https://github.com/playframework</url>
        </developer>
    </developers>
    <dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-guice_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>build-link_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>routes-compiler_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>sbt-routes-compiler_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-akka-http-server_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-akka-http2-support_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-cache_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-ehcache_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-caffeine-cache_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-jcache_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-jdbc-api_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-jdbc_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-jdbc-evolutions_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-java_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-java-forms_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-joda-forms_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-java-jdbc_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-java-jpa_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-netty-server_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-server_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-logback_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-ws_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-ahc-ws_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-openid_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>run-support_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>sbt-plugin_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>sbt-scripted-tools_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-specs2_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-test_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-exceptions_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>filters-helpers_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-streams_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-cluster-sharding_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
            <dependency>
                <groupId>com.typesafe.play</groupId>
                <artifactId>play-java-cluster-sharding_2.12</artifactId>
                <version>2.8.1+749-78742811</version>
            </dependency>
        </dependencies>
    </dependencyManagement>
    <properties>
        <info.apiURL>https://www.playframework.com/documentation/latest/api/scala/index.html</info.apiURL>
    </properties>
</project>
```

</details>

File: build.sbt [hunk 1]
```diff
@@ -58,6 +58,66 @@ lazy val StreamsProject = PlayCrossBuiltProject("Play-Streams", "core/play-strea
...
+    autoScalaLibrary := false,
+    crossVersion := CrossVersion.disabled, // this setting removes the scala bin version from the artifact name
+    crossScalaVersions := Seq(interplay.ScalaVersions.scala212),
+    scalaVersion := interplay.ScalaVersions.scala212,
+    crossPaths := false,
+    autoScalaLibrary := false,
+    pomExtra := pomExtra.value :+ {
+      val playDeps = Def.settingDyn {
+        // all Play artifacts are cross compiled
+        // FIXME: what about Play-Java-JPA, Play-Java-JDBC, Play-exceptions??
```
Comment (ignasi35) [inline build.sbt:86]: TODO: review what happens with artifacts that are not cross-compiled.

**Comment by ihostage** [comment]
@ignasi35 Don't Play-BOM should depend on Akka-BOM? 🤔

**Comment by ignasi35** [comment]
> @ignasi35 Don't Play-BOM should depend on Akka-BOM? 🤔

Hmmm, good point.

I do know we should not add all the dependencies in the Play BOM like we did on the Lagom BOM, though. You mentioned one BOM _depending_ on the other and it may be a good idea.

**Comment by ihostage** [comment]
> I do know we should not add all the dependencies in the Play BOM like we did on the Lagom BOM, though.

Yes, I absolutely agree with this point. 

> You mentioned one BOM depending on the other and it may be a good idea.

When I thinking about BOM for Akka/Play/Lagom, I every think about transitive dependencies between BOMs. Otherwise for explicit using several BOMs user needs to know Akka-Play-Lagom versions compatibility and It's look like a dangerous way for users.

**Comment by ihostage** [comment]
For example [spring-boot-dependencies.pom](https://repo1.maven.org/maven2/org/springframework/boot/spring-boot-dependencies/2.4.0/spring-boot-dependencies-2.4.0.pom) transitively depends on `java-driver-bom` (Cassandra), `metrics-bom` (Dropwizard), `jackson-bom`, `spring-data-bom`, `spring-framework-bom`, etc.
I think that Akka/Play/Lagom approach can be similar 😄

**Comment by ignasi35** [comment]
So we would add:

```xml
<dependency>
	<groupId>com.typesafe.akka</groupId>
	<artifactId>akka-bom</artifactId>
	<version>${akka.version}</version>
	<type>pom</type>
	<scope>import</scope>
</dependency>
<dependency>
	<groupId>com.typesafe.akka</groupId>
	<artifactId>akka-http-bom</artifactId>
	<version>${akka-http.version}</version>
	<type>pom</type>
	<scope>import</scope>
</dependency>
```

Sounds doable. The only issue I see is that we'd have to hold this improvement until both Akka and Akka-HTTP have released a version that includes a BOM. 

I think I'd rather get this PR forward and create a new issue with the "Depend on Akka BOMs from the Play BOM" improvement to work on it in the future.

**Comment by ignasi35** [comment]
@ihostage I think we must not add a dependency from the Play BOM to the Akka BOM. I've been investigating more and I think there are two usages of Maven's `<type>pom</type>` artifacts. See https://repo1.maven.org/maven2/org/springframework/spring-framework-bom/5.3.1/spring-framework-bom-5.3.1.pom for a counter-example.

The way I see it:

1. A `foo-bom` artifact would only list artifacts it controls. As an example, a `play-bom` only lists play artifacts. 
2. A `bar-dependencies` is produced by framework mantainers to align all the dependencies. Lagom's `lagom-maven-dependencies` lists lagom artifacts, and akka artifacts, and jackson, and... 

So, this PR is creating a `play-bom`. On a separate issue we could discuss the convenience of having a `play-dependencies` pom that would include Akka artifacts, but also a huge list of other artifacts.

Does that make sense?

**Comment by ihostage** [comment]
> Does that make sense?

@ignasi35 Yes. I thought about splitting these cases into different artifacts too. 😄 
```
akka-microservices-dependencies
   |--- akka-bom   
   |--- akka-http-bom
   |--- akka-persistence-bom
   |--- jackson-bom
   |--- etc
```

**Review by ignasi35 — COMMENTED** [review]
LGTM (can't approve because I was the original submitter of the PR)

**Comment by ignasi35** [comment]
@mergifyio backport 2.8.x

**Pull request merged**